Mathlib/LinearAlgebra/TensorAlgebra/Hopf.lean

```diff
@@ -0,0 +1,210 @@
+/-
+Copyright (c) 2025 Nikolas Tapia. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nikolas Tapia
+-/
+module
+
+public import Mathlib.LinearAlgebra.TensorAlgebra.Basic
+public import Mathlib.RingTheory.Coalgebra.Convolution
+public import Mathlib.RingTheory.HopfAlgebra.Basic
+
+/-!
+# Hopf algebra structure on `TensorAlgebra R M`
+
+In this file we implement the natura Hopf algebra structure on `TensorAlgebra R M`.
+The comultiplication is the unique algebra map satisfying `comul ((ι R) x) = (ι R) x ⊗ 1 + 1 ⊗ (ι R)
+x` for all `x : M`.
+`algebraMapInv` acts as the counit, and the antipode is the unique algebra map `antipode :
+TensorAlgebra R M → (TensorAlgebra R M)ᵐᵒᵖ` induced by `fun x => op -(ι R) x`.
+-/
+@[expose] public section
+
+namespace TensorAlgebra
+open scoped TensorProduct RingTheory.LinearMap Coalgebra
+open LinearMap TensorProduct
+
+variable (R : Type*) [CommRing R] {M : Type*} [AddCommMonoid M] [Module R M]
+
+local notation "T["M"]" => TensorAlgebra R M
+
+/-- Linear map inducing the comultiplication in `TensorAlgebra R M`. -/
+def comul' : M →ₗ[R] T[M] ⊗[R] T[M] :=
+  ((ι R) ⊗ₘ (Algebra.linearMap R T[M])) ∘ₗ (TensorProduct.rid R M).symm +
+  ((Algebra.linearMap R T[M]) ⊗ₘ (ι R)) ∘ₗ (TensorProduct.lid R M).symm
+
+/-- Comultiplication in `TensorAlgebra R M` as an algebra map. -/
+def comul : T[M] →ₐ[R] T[M] ⊗[R] T[M] := lift R (comul' R)
+
+/-- Antipode in `TensorAlgebra R M` as an algebra map. -/
+def antipode : T[M] →ₗ[R] T[M] := (MulOpposite.opLinearEquiv R).symm.comp
+  (lift R ((MulOpposite.opLinearEquiv R).comp (-(ι R)))).toLinearMap
+
+@[simp]
+lemma comul_apply (x : M) : comul R (ι R x) = ι R x ⊗ₜ[R] ↑1 + ↑1 ⊗ₜ[R] ι R x := by
+  unfold comul comul'
+  simp
+
+@[simp]
+lemma counit_ι_apply (x : M) : algebraMapInv ((ι R) x) = 0 := by
+  rw [algebraMapInv]
+  simp
+
+@[simp]
+lemma counit_ι_eq_zero : algebraMapInv.toLinearMap ∘ₗ (ι R) = (0 : M →ₗ[R] R) := by
+  ext x
+  simp
+
+theorem rTensor : (Algebra.TensorProduct.map algebraMapInv (AlgHom.id R T[M])).comp (comul R) =
+    (Algebra.TensorProduct.lid R T[M]).symm.toAlgHom := by
+  unfold comul comul'
+  ext x
+  rw [algebraMapInv]
+  simp
+
+theorem lTensor : (Algebra.TensorProduct.map (AlgHom.id R T[M]) algebraMapInv).comp (comul R) =
+    ↑(Algebra.TensorProduct.rid R R T[M]).symm := by
+  unfold comul comul'
+  ext x
+  rw [algebraMapInv]
+  simp
+
+@[simp]
+lemma antipode_ι_apply (x : M) : antipode R ((ι R) x) = -(ι R) x := by
```

Comment:
If this is true, doesn't it mean that the antipode is just negation? In this case, it would just be [LinearEquiv.neg](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Equiv/Basic.html#LinearEquiv.neg).

Reply:
This holds only on the image of ι R. For example
```lean4
antipode R (ι R x * ι R y) = ι R y * ι R x
```
In general, for x in T[M]
```
antipode R ((tprod R M n) x) = (-1)^n • (tprod R M n) (fun i => x (n - i))
```
so there's an alternating minus sings and a reversal. I'm not sure if this is the best way to write it.